File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find input value =
    let sorted = Array.sort input
    let rec search input indexOffset value = 
        match input with 
        | [||] -> None
        | _ -> 
            let mid = input.Length / 2
            match input.[mid] with
            | x when x > value -> 
                let right = Array.take mid input
                search right indexOffset value
            | x when x < value -> 
                let left = Array.skip (mid + 1) input
                search left (indexOffset + mid + 1) value
            | _ -> indexOffset + mid |> Some
    search sorted 0 value

File: go/binary-search/binary_search.go
package binarysearch

func SearchInts(sorted []int, target int) int {
	if len(sorted) == 0 {
		return -1
	}
	mid := len(sorted) / 2
	if target == sorted[mid] {
		return mid
	}
	if target < sorted[mid] {
		return SearchInts(sorted[:mid], target)
	}
	higher := SearchInts(sorted[mid+1:], target)
	if higher == -1 {
		return -1
	}
	return mid + 1 + higher
}
